[CORL-1655] Show trace ID's to resolve client error's easier

nick-funk opened the pull request:
## What does this PR do?

Creates a unique trace ID on every server request that can be attached to error messages, logged in the server logs, and shown to the user. The trace ID is oblique in that it tells the user nothing of what caused the error, but is an identifier they can provide to support staff that can easily be correlated to server logs to assist in resolving/debugging the issue.

## What changes to the GraphQL/Database Schema does this PR introduce?

None

## How do I test this PR?

- Throw an error somewhere in the code base
  - An easy one is to make the `body` resolver throw an error in `src/core/server/graph/resolvers/Comment.ts` (Ln: 56)
- Visit the site and trigger this error
- Should see a call out with the trace ID
- Check console running Coral and find matching `contextID` for the error
- Test this in both development and production builds

Example:
<img width="462" alt="image" src="https://user-images.githubusercontent.com/5751504/122956375-2e1cbb00-d33e-11eb-8291-4a3600401d50.png">

nick-funk committed:
Create traceID on request and graph context

nick-funk committed:
Show traceID's for GraphQL query errors client side

nick-funk committed:
Update all query handlers to user QueryError

nick-funk committed:
Make traceID's come through on production builds

wyattjoh commented on src/core/server/app/middleware/graphql/apolloServer.ts:
```diff
@@ -31,6 +32,7 @@ function contextProvider(options: ContextProviderOptions) {
       tenant,
       logger,
       site,
+      traceID: uuidv4(),
```
The `id` on this already is sent as the Trace ID in the `X-Trace-ID` header of the response.

wyattjoh commented on src/core/server/app/middleware/graphql/apolloServer.ts:
```diff
@@ -45,6 +47,9 @@ function contextProvider(options: ContextProviderOptions) {
 
+    // Set trace ID for debug events if we error
+    req.traceID = opts.traceID;
+
```
This shouldn't be duplicated as it's already available on the request as `req.coral.id`.

wyattjoh commented on src/core/server/graph/context.ts:
```diff
@@ -116,6 +118,7 @@ export default class GraphContext {
     this.signingConfig = options.signingConfig;
     this.clientID = options.clientID;
     this.reporter = options.reporter;
+    this.traceID = options.traceID || "";
```
This is already provided as `options.id` on the GraphContext.

wyattjoh commented on src/core/server/graph/plugins/helpers.ts:
```diff
@@ -108,7 +108,13 @@ function hoistCoralErrorExtensions(
+  rawError.traceID = ctx.traceID;
+
+  if (err.extensions) {
+    err.extensions.traceID = ctx.traceID;
```
Should be able to use `ctx.id` instead.

nick-funk committed:
Update snapshot to account for QueryError component

wyattjoh commented on src/core/server/graph/plugins/logger.ts:
```diff
@@ -45,7 +45,10 @@ export function logAndReportError(
+  ctx.logger.error(
+    { err, report, traceID: ctx.traceID },
+    "graphql query error"
+  );
```
Shouldn't need to add the `traceID` here, as `ctx.logger` already adds a `contextID` which is the same ID.

wyattjoh reviewed (changes requested):
Looks good! I'd just suggest that you should use the already existing trace ID rather than creating a second one. Happy to discuss!

nick-funk committed:
Use existing GraphContext id for traceID

wyattjoh commented on src/core/server/types/express.ts:
```diff
@@ -18,13 +18,15 @@ export interface CoralRequest {
 export interface Request<T = CoralRequest> extends ExpressRequest {
   coral: T;
   user?: User;
+  traceID?: string;
```
These changes aren't needed now.

wyattjoh commented on src/core/server/app/middleware/error.ts:
```diff
@@ -126,6 +134,7 @@ function wrapAndReport(
     context: {
       error,
       report,
+      traceID: req.traceID,
```
These should use `req.coral.id` instead of `req.traceID`

wyattjoh reviewed (approved):
Server integration and client side integration looks nice! Excited to see Vinh's comment as well!

alinhle reviewed (changes requested):
Exact review is coming.

alinhle commented on src/core/client/framework/lib/errors/unknownServerError.ts:
```diff
@@ -12,6 +12,7 @@ interface UnknownErrorExtension {
 export default class UnknownServerError extends Error {
```
`UnknownServerError`, `InvalidRequestError`, etc should inherit from a new class `TraceableError`.

```ts
class TraceableError extends error {
  public readonly traceID: string;
}
```

alinhle commented on src/core/client/framework/lib/network/extractGraphQLError.ts:
```diff
@@ -12,19 +12,48 @@ function isGraphQLErrorWithExtension(
     // TODO: (cvle) harmonize with server.
-    return null;
+    return new MultiError(errors);
   }
```
I just ran a thorough analysis and these lines actually should never happen.
When GraphQL Standard error occurs they always return a non 200 status code and as such will never hit this routine.

So we can replace line 47-57 with:
```ts
if (errors.length > 1 || !isGraphQLErrorWithExtension(errors[0])) {
  return new Error(`Unexpected errors: ${JSON.stringify(errors}`);
}
```

`MultiError` can then be removed.

nick-funk replied:
That's odd, I was easily able to make these happen with a simple error on a GraphQL resolver server side. Particularly common in dev builds.

nick-funk replied:
I removed it and things keep on chugging, going to keep it this way until I find a reason to re-add a `MultiError` or similar.

alinhle replied:
> That's odd, I was easily able to make these happen with a simple error on a GraphQL resolver server side. Particularly common in dev builds.

Oh if you can provide an example, it would be super helpful!

alinhle replied:
You were right, `MultiError` does exist 😱 I missed that. We can sync about this later.

alinhle commented on src/core/server/graph/plugins/helpers.ts:
```diff
@@ -108,7 +108,13 @@ function hoistCoralErrorExtensions(
+
+  if (err.extensions) {
+    err.extensions.traceID = ctx.id;
+  }
```
We don't need to attach the traceID to the error directly. Extensions are always available so we can rely on that. This will be more spec compilant.

Should put implementation in `serializeExtensions` e.g. in line 107 it should say 
```ts
const extensions = originalError.serializeExtensions(bundle, traceID);
```

wyattjoh replied:
That actually makes a lot of sense!

nick-funk replied:
This does not work. Also tried modifying the `serializeExtensions` method to receive the traceID and set it in the returned object, but that also does not work. Am fairly certain I need to set this here.

alinhle replied:
Hmm I just tried it and it worked, let's chat about this.

alinhle commented on src/core/server/services/comments/pipeline/phases/index.ts:
```diff
@@ -61,9 +61,9 @@ export const moderationPhases: IntermediateModerationPhase[] = [
   detectLinks,
 
   // Apply any pre-existing conditions to these comments.
+  statusPreModerateNewCommenter,
```
This seems to be from another PR that I just reviewed?

alinhle commented on src/core/client/framework/lib/errors/unknownServerError.ts:
```diff
@@ -21,6 +22,17 @@ export default class UnknownServerError extends Error {
 
+    const traceID: string | null | undefined = (error as any).traceID as string;
+    const traceIDs: string[] | null | undefined = (error as any)
+      .traceIDs as string[];
```
As suggested `traceIDs` doesn't need to be handled, as we can remove `MultiError` class.

nick-funk replied:
Please see previous comment for why I created MultiError.

alinhle commented on src/core/client/framework/lib/errors/unknownServerError.ts:
```diff
@@ -21,6 +22,17 @@ export default class UnknownServerError extends Error {
       Error.captureStackTrace(this, UnknownServerError);
     }
 
+    const traceID: string | null | undefined = (error as any).traceID as string;
```
Instead of casting `error as any` add `traceID` to  type `UnknownErrorExtension` on line 4.

alinhle commented on src/core/client/ui/components/v3/QueryError/QueryError.tsx:
```diff
@@ -0,0 +1,58 @@
+    }
+  } catch {
+    return <div>{error.message}</div>;
+  }
```
When following the other suggesions, we can use the common `TraceableError` class and also provide the same nice ui for errors without the traceID:
```suggestion
  let traceID = error instanceof TraceableError ? error.traceID : "";
  return (
    <CallOut
      color="error"
      title={
        <Localized id="common-error-title">
          <span>An error has occurred</span>
        </Localized>
      }
    >
      {traceID && (
        <>
          <Localized id="common-error-traceID">
            <div className={styles.heading}>Trace ID</div>
          </Localized>
          <div className={styles.section}>{traceID}</div>
        </>
      )}
      <Localized id="common-error-message">
        <div className={styles.heading}>Message</div>
      </Localized>
      <div className={styles.section}>{message}</div>
    </CallOut>
  );
```

nick-funk replied:
Done!

alinhle commented on src/core/client/framework/lib/errors/invalidRequestError.ts:
```diff
@@ -35,9 +36,11 @@ export default class InvalidRequestError extends Error
+    this.traceID = (extension as any).traceID;
     this.param = extension.param;
     this.message = extension.message || extension.code;
   }
```
See review of UnknownServerError

nick-funk committed:
Create TraceableError and extend it through other errors

nick-funk committed:
Handle traceID in unknownServerError

nick-funk committed:
Remove MultiError and update error handling

nick-funk committed:
Update invalidRequestErrorExtension to use traceID

alinhle commented on src/core/client/framework/lib/errors/invalidRequestError.ts:
```diff
@@ -12,18 +14,20 @@ interface InvalidRequestExtension {
   message?: string;
   id?: string;
   param?: string;
+  traceID?: string;
```
This is a required field and is always provided.

alinhle commented on src/core/client/framework/lib/errors/invalidRequestError.ts:
```diff
@@ -12,18 +14,20 @@ interface InvalidRequestExtension {
   public readonly extension: InvalidRequestExtension;
   public readonly code: ERROR_CODES;
   public readonly id?: string;
+  public readonly traceID?: string;
```
This is a required field and is always provided.

alinhle commented on src/core/client/framework/lib/errors/networkOfflineError.ts:
```diff
@@ -1,8 +1,10 @@
  * but network was offline.
  */
-export default class NetworkOfflineError extends Error {
+export default class NetworkOfflineError extends TraceableError {
```
This is not a TraceableError.

alinhle commented on src/core/client/framework/lib/errors/relayNetworkRequestError.ts:
```diff
@@ -1,14 +1,17 @@
  * RelayNetworkRequestError wraps Request errors thrown by Relay Network Layer.
  */
-export default class RelayNetworkRequestError extends Error {
+export default class RelayNetworkRequestError extends TraceableError {
```
This is also not a TraceableError.

alinhle commented on src/core/client/framework/lib/errors/unknownServerError.ts:
```diff
@@ -1,17 +1,21 @@
+export default class UnknownServerError extends TraceableError {
   // Keep origin of original server response.
   public origin: UnknownErrorExtension;
+  public traceID?: string;
```
This is a required field and is always provided.

alinhle commented on src/core/client/framework/lib/errors/unknownServerError.ts:
```diff
@@ -21,6 +25,12 @@ export default class UnknownServerError extends Error {
+    if (traceID) {
+      this.traceID = traceID;
+    }
+
```
`this.traceID = error.traceID` should suffice

alinhle commented on src/core/client/framework/lib/errors/unknownServerError.ts:
```diff
@@ -1,17 +1,21 @@
  */
 interface UnknownErrorExtension {
   code: string;
+  traceID?: string;
```
This is a required field and is always provided.

alinhle commented on src/core/client/framework/lib/errors/traceableError.ts:
```diff
@@ -0,0 +1,5 @@
+class TraceableError extends Error {
+  public readonly traceID?: string;
```
This should be a required field.

alinhle commented on src/core/client/ui/components/v3/QueryError/QueryError.tsx:
```diff
@@ -0,0 +1,56 @@
+  } catch {
+    return <div>{error.message}</div>;
+  }
+};
```
I suggest rendering the same or similar nice UI no matter if traceID is available or not, see: https://github.com/coralproject/talk/pull/3593#discussion_r657894869.

nick-funk committed:
Set traceID on the error extensions

Also some other cleanup making traceID's mandatory
on trace-able error objects.

nick-funk committed:
Simplify client side error handling for trace ID's

nick-funk committed:
Remove faked out error for debugging purposes

nick-funk committed:
Pull traceID from request context instead of response

nick-funk committed:
Make TraceableError use a required constructor

nick-funk committed:
Re-introduce multi-error for handling compound graphQL errors

nick-funk committed:
Simplify QueryError and always return nice CallOut

alinhle committed:
chore: refactor

kodiakhq[bot] committed:
Merge branch 'develop' into feat/CORL-1655-trace-id
